### ISSUE TITLE
Feature Request: Train lora

### ISSUE THREAD

**Issue by kirel** [filipstrand/mflux#78]
Docs say
> actual training support is coming

— is there any work visible somewhere?

**Comment by anthonywu** [comment]
@kirel follow this adjacent work: https://github.com/ml-explore/mlx-examples/tree/main/flux#training-example - this project is upstream from the MLX team, and I think mflux could/should just accept the trained Loras from that project

**Comment by filipstrand** [comment]
As Anthony suggested, right now I would also recommend the mlx-examples project if you want to run fine-tuning locally. I quickly tried loading a LoRA from mlx-examples, and can confirm that it does not work with mflux currently. Inspecting the weight matrices, there seem to be some other naming conventions. But I think a renaming and reshaping can fix this. For example,  

```
state_dict["lora_unet_single_blocks_37_linear2.lora_down.weight"] = state_dict["single_blocks.37.linear2.lora_a"].transpose(0, 1)
```

I have not actually tried this yet as my one fine-tuning attempt failed (probably my fault for not training with enough iterations).  But I can at least confirm that the shapes seem to match and the LoRA does load in mflux after the renaming and reshaping. If anyone could point me to a successful LoRA that was fine-tuned from mlx-examples, it would be really nice as a test case to confirm that the translation is working as expected. 

Regarding fine-tuning support in mflux - I have personally not found the time lately to really dig into this feature as I first planned (mflux is still a nights and weekends project for me, but I would have loved to be able to spend more time on it :) ). My hope is to eventually find the time to implement this, and, of course, help with reviewing PRs if anyone else contributes a working implementation.

**Comment by filipstrand** [comment]
@azrahello No, quite the opposite, great to have a beta tester for the upcoming feature :). 

There is not that much left to do on the open Dreambooth PR, but I currently have an issue with training not being completely deterministic when trying to resume from an existing checkpoint. This is a minor detail, but I would like to get it right before releasing the feature. Hopefully, this will be resolved soon! 

And thanks for your generosity! I am limited to a more basic setup, so my training runs are probably quite a bit slower than yours :). Maybe once I have the PR in a camera-ready state, if you want/have the time, I might ask you to do some test runs. I'll keep in touch! 

I was also thinking that we could have a benchmark (like we do with regular image generation) of how long it takes to complete a mini-training run, say 20 gradient updates based on the default dataset in the repo. I know I would find it interesting to compare the numbers. 

Also, if you have any other feedback on the interface, config files etc, please feel free to share :)

**Comment by azrahello** [comment]
Yes, I do it gladly. Curiosity drives me, and it helps me understand more things, in addition to purely academic reasons. :).
If I understood correctly, this is what you asked me:
python train.py --model dev --quantize 4 --train-config
359.38s user, 185.91s system, 51% CPU, 17:39.17 total.
[0000100_loss.pdf](https://github.com/user-attachments/files/18030646/0000100_loss.pdf)

**Comment by filipstrand** [comment]
Thanks for your feedback! 

Since there is so much configuration needed for training, my idea here was that you would always **need** to send in a training configuration file (i.e something like `--train-config "src/mflux/dreambooth/_example/train.json"`. From there we will read all the necessary things, such as model choice, seed,  number of steps etc. In other words, all other parameters that a user might send in will be ignored - we should probably add some validation that says this (The fact that you currently have to send in `--model` also will be removed, since we already have that in the config). 

When you say ` I tried using LoRAs, but they had no effect`, were you sending in an existing LoRA adapter to the training script? If so, then it is reasonable that this did not work as you said. However, I should add support for this since it feels like a good use case, but I think it could slightly more complicated due to the fact that pre-loaded LoRAs could be of different shapes than the one being trained and so on... so won't be in the initial release, but definitely a cool feature to support!). 

Regarding optimiser vs adapter: Adapter would be the final LoRA that you train, i.e the model weights. The optimiser is the current optimiser state - since Adam / AdamW changes over time, my idea was to also save these to disk so that if you would like to resume a training run at a later time, we could pick up **exactly** where we left off - which is a nice feature to have. I have a little problem with this at the moment, but will try to fix it this weekend. The idea is that when you resume training from a checkpoint by `--train-checkpoint "0000001_checkpoint.zip"`, MFLUX will read in the full state of the training at that step (`0000001`) and resume from there. 
  
Regarding the benchmarks, thanks for your numbers! I have not decided fully yet, but was planning on having a `train_benchmark.json` file or something similar in the repo the we could run and hardcode all values (maybe have 
```json
"training_loop": {
        "num_epochs": 20,
        "batch_size": 1
    },
```
for 100 steps over the 5 images, and maybe reduce the image resolution a bit so that it takes less time (just to get a rough idea) 

```json
 "width": 128,
 "height": 128,
```

For example if I ran this on my 32GB M2 Max, and looked at `0000100_checkpoint/checkpoint.json` I got 

```
"metadata": {
         ...
        "duration": "13 minutes and 50 seconds",
        "number_of_training_examples": 5
    },
```

**Comment by azrahello** [comment]
I agree with the idea of directly sending the configuration file, and it would also be great to have the option to create it interactively. The values I entered in the configuration file are correct, but perhaps my environment is a bit “messy.” I’m currently running another training session in a clean environment that I just created.

The values I entered were correct. After training, I tried using the adapter file with mflux-comfyui as a LoRA. I did two generations, one with and one without it, and the output was identical with a weight of 1.0.

```
{
    "metadata": {
        "start": "2024-12-07 11:29:46",
        "end": "2024-12-07 11:47:31",
        "duration": "17 minutes and 44 seconds",
        "number_of_training_examples": 5
    },
  
```

```
{
    "model": "dev",
    "seed": 42,
    "steps": 20,
    "guidance": 3.0,
    "quantize": 4,
    "width": 128,
    "height": 128,
    "training_loop": {
        "num_epochs": 20,
        "batch_size": 1
    },
```
Isn’t it strange that it takes me longer?

**Comment by filipstrand** [comment]
Okay thanks for confirming. Indeed strange that your training run takes longer than mine... Will have to investigate this. I am a bit short on time for the remainder of today, but I could update the main branch `config/train.json` file so that we are sure we are running the same file (even though your settings looks correct here). I have pushed that now, changed it to 10 epochs just to have it finish a bit quicker (50 grad steps in total) 

And speaking of reading in the lora, that makes sense since I have not fixed that yet. Will address this shortly!  

But I am happy to report that the resuming from training is indeed fully deterministic and I have a test to confirm this now (turns out it was before as well, but I accidentally looked at the wrong stats that made me think otherwise.)

**Comment by azrahello** [comment]
Yesterday, I ran additional tests. Before starting, I deleted all environments and then created a new one with Python 3.12:

conda create -n test python=3.12 -y  
conda activate test  

Then, in the mflux directory, I ran:

pip install -r requirements.txt  

I executed the training using the default configuration file without making any changes and obtained the following results:

```
{
    "metadata": {
        "start": "2024-12-08 08:49:03",
        "end": "2024-12-08 08:53:54",
        "duration": "4 minutes and 51 seconds",
        "number_of_training_examples": 5
    },
    "files": {
        "config": "0000050_config.json",
        "optimizer": "0000050_optimizer.safetensors",
        "lora_adapter": "0000050_adapter.safetensors",
        "iterator": "0000050_iterator.json",
        "loss": "0000050_loss.json"
    }
}
```

The strange thing is that the training speed was around 5 seconds per iteration (s/it), which seems reasonable. Immediately after that, without changing anything, I started another training session with a different configuration file
[0000050_config.json](https://github.com/user-attachments/files/18051289/0000050_config.json)
From the beginning, the iterations were slower, averaging around 7 s/it (starting from 12 and stabilizing at 7). With this config file obtained 
```
{
    "metadata": {
        "start": "2024-12-08 09:20:17",
        "end": "2024-12-08 09:27:09",
        "duration": "6 minutes and 52 seconds",
        "number_of_training_examples": 5
    },
    "files": {
        "config": "0000050_config.json",
        "optimizer": "0000050_optimizer.safetensors",
        "lora_adapter": "0000050_adapter.safetensors",
        "iterator": "0000050_iterator.json",
        "loss": "0000050_loss.json"
    }
}

```

Thinking it might be due to some memory issue, I restarted the default training again. Once again, it ran at 5 s/it. It’s a mystery to me.

**Comment by filipstrand** [comment]
Interesting... memory issue sounds plausible. On one of my machines I  also got the same duration of `"duration": "6 minutes and 52 seconds",`. I'll perhaps look more into this once the feature is complete and fully works as intended.  Yesterday, I've made some updates so that we can read in the MFLUX-trained LoRAs the typical way, but there are a few things to iron out here too and possibly some refactoring and testing that needs to be done to ensure things work as expected here.

**Comment by azrahello** [comment]
Yesterday, I ran a couple more tests and managed to complete 4100 steps in about an hour, training on a dataset of around 41 images. Today, I repeated the test using the default training file, changing only the number of plots from 1 to 10  
```
{
    "metadata": {
        "start": "2024-12-10 08:21:43",
        "end": "2024-12-10 08:24:02",
        "duration": "2 minutes and 18 seconds",
        "number_of_training_examples": 5
    },
    "files": {
        "config": "0000050_config.json",
        "optimizer": "0000050_optimizer.safetensors",
        "lora_adapter": "0000050_adapter.safetensors",
        "iterator": "0000050_iterator.json",
        "loss": "0000050_loss.json"
    }
}
```

The problem is that verbosity somehow goes out the window because it’s hard to understand what it’s doing at certain moments. It seems to freeze, but in reality, it’s still training… and doing a pretty decent job at that.

**Comment by filipstrand** [comment]
Closing this PR as this feature is now merged to main and included in v.0.5.0
